[logging] Log individual Triton kernel compilation times to dynamo_compile

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147022

Summary: Gather the compilation time of individual triton kernels and log them to dynamo_compile:
* Time compilation in `_worker_compile_triton` and pass back to the main process and logged from `get_result()`.
* Added a way to track the "top N" (or N most-expensive compiles) in the metrics_context. I did this because I doubt we really care to capture potentially thousands of kernel compile times. That would be problematic for scuba logging anyway, so let's limit the number we track from the beginning. Arbitrarily chose 25 for now.
* Format the list of compile times as a json string before logging.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/nc4dzm3r

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

masnesral committed:
[logging] Log individual Triton kernel compilation times to dynamo_compile / pt2_compile_events

Summary:

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Log individual Triton kernel compilation times to dynamo_compile"


Summary:

Test Plan:

[ghstack-poisoned]

masnesral commented on torch/_inductor/async_compile.py:
```diff
@@ -324,9 +329,15 @@ def get_result() -> CachingAutotuner:
                 dynamo_compile_column_us="triton_compile_time_us",
                 log_waitcounter=True,
             ):
+                start_ns = time_ns()
```
This part is a little weird because we're already inside a dynamo_timed and yet tracking the elapsed time again. In the common case, dynamo_timed is going to add the timing to the running total. But in this case, I want to add the individual compile time to the running list of the top N. I could add this ability to dynamo_timed, but since this is the only instance, I didn't love complicating dynamo_timed even more.

masnesral commented on torch/_inductor/async_compile.py:
```diff
@@ -393,42 +404,51 @@ def halide(self, meta: HalideMeta, source_code: str):
-                        pbar.update(1)
-
-            _compile_end()
+        if get_compile_threads() > 1:
```
The indentation was getting too much for me, so added `_wait_futures`

masnesral commented on torch/_inductor/async_compile.py:
```diff
@@ -393,42 +404,51 @@ def halide(self, meta: HalideMeta, source_code: str):
+                pbar.set_postfix_str(key)
+            try:
+                scope[key] = result.result()
+                if isinstance(result, LambdaFuture) and result.elapsed_us:
```
Don't love this check for the future type. Open to suggestions if you hate it...

jamesjwu replied:
Do you know `elapsed_us` is an attribute on LambdaFuture? That's only true for LambdaFutures generated by async_compile.triton, right? 

You could write a new Future type called TritonLambdaFuture if you'd like. But don't really think it's a big deal overall, you're just basically having the LambdaFuture return a tuple.

masnesral committed:
Update on "[logging] Log individual Triton kernel compilation times to dynamo_compile"


Summary: Gather the compilation time of individual triton kernels and log them to dynamo_compile:
* Time compilation in `_worker_compile_triton` and pass back to the main process.
* Added a `timed` param to `LambdaFuture` that indicates the future's `result()` method will additionally provide an elapsed time.
* Added a way to track the "top N" (or N most-expensive compiles) in the metrics_context. I did this because I doubt we really care to capture potentially thousands of kernel compile times. That would be impractical for scuba logging anyway, so let's limit the number we track from the beginning.
* Format the list of compile times as a json string before logging.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/9ih7i55x

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Log individual Triton kernel compilation times to dynamo_compile"


Summary: Gather the compilation time of individual triton kernels and log them to dynamo_compile:
* Time compilation in `_worker_compile_triton` and pass back to the main process.
* Added a `timed` param to `LambdaFuture` that indicates the future's `result()` method will additionally provide an elapsed time.
* Added a way to track the "top N" (or N most-expensive compiles) in the metrics_context. I did this because I doubt we really care to capture potentially thousands of kernel compile times. That would be impractical for scuba logging anyway, so let's limit the number we track from the beginning.
* Format the list of compile times as a json string before logging.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/9ih7i55x

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Log individual Triton kernel compilation times to dynamo_compile"


Summary: Gather the compilation time of individual triton kernels and log them to dynamo_compile:
* Time compilation in `_worker_compile_triton` and pass back to the main process.
* Added a `timed` param to `LambdaFuture` that indicates the future's `result()` method will additionally provide an elapsed time.
* Added a way to track the "top N" (or N most-expensive compiles) in the metrics_context. I did this because I doubt we really care to capture potentially thousands of kernel compile times. That would be impractical for scuba logging anyway, so let's limit the number we track from the beginning.
* Format the list of compile times as a json string before logging.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/9ih7i55x

[ghstack-poisoned]

jamesjwu commented on torch/_inductor/async_compile.py:
```diff
@@ -307,14 +312,14 @@ def reload_kernel_in_parent():
                 kernel.precompile(
                     warm_cache_only=False, reload_kernel=reload_kernel_in_parent
                 )
-                return kernel
```
This `get_result()` function gets called on the parent class. So you already have the elapsed_us here, in triton specific lambda. Can you not call get_metrics_context().add_top_N() here instead of at the wait? Then you wouldn't need to check a condition of LambdaFuture in async_compile.wait. 

Is it slower to put it here for some reason? Doesn't seem like it would be to me.

masnesral replied:
Yeah, great suggestion. I'll do that instead.

jamesjwu reviewed (approved):
The generalization of LambdaFuture to have a specific timed field is nice, but it doesn't translate perfectly to the fact that you take the elapsed_us field of each LambdaFuture and put it into a triton specific metric. It kind of presupposes that all/only triton LambdaFutures are timed. 

Approving to unblock, but see suggestion below

masnesral committed:
Update on "[logging] Log individual Triton kernel compilation times to dynamo_compile"


Summary: Gather the compilation time of individual triton kernels and log them to dynamo_compile:
* Time compilation in `_worker_compile_triton` and pass back to the main process.
* Added a `timed` param to `LambdaFuture` that indicates the future's `result()` method will additionally provide an elapsed time.
* Added a way to track the "top N" (or N most-expensive compiles) in the metrics_context. I did this because I doubt we really care to capture potentially thousands of kernel compile times. That would be impractical for scuba logging anyway, so let's limit the number we track from the beginning.
* Format the list of compile times as a json string before logging.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/nc4dzm3r

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
